File: solutions/343/main.py
from typing import List 

class Solution:
    def integerBreak(self, n: int) -> int:
        dp = [0] * (n+1)
        dp[2] = 1
        for i in range(3, n+1):
            for j in range(1, i-1):
                dp[i] = max(dp[i], max(dp[i-j]*j, (i-j)*j))
        return dp[n]

if __name__ == "__main__":
    s = Solution()
    n = 10 
    assert s.integerBreak(n) == 36

File: solutions/343/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Thu Nov 11 10:03:09 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int integerBreak(int n) {
        vector<int> dp(n+1);
        // initialize
        dp[2] = 1;
        for (int i = 3; i <= n; ++i) {
            for (int j = 1; j < i-1; ++j) {
                dp[i] = max(dp[i], max(j * (i-j), j * dp[i-j]));
            }
        }
        return dp[n];
    }
};

TEST(Test343, SimpleTest) {
    const int n = 10; 
    Solution s; 
    EXPECT_EQ(s.integerBreak(n), 36);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
